source/opt/aggressive_dead_code_elim_pass.cpp

```diff
@@ -826,9 +841,19 @@ bool AggressiveDCEPass::ProcessGlobalValues() {
   return modified;
 }
 
-AggressiveDCEPass::AggressiveDCEPass() = default;
+AggressiveDCEPass::AggressiveDCEPass(bool preserve_bindings,
+                                     bool preserve_spec_constants) :
+                                     preserve_bindings_(preserve_bindings),
+                                     preserve_spec_constants_(preserve_spec_constants) {}
 
 Pass::Status AggressiveDCEPass::Process() {
+  // Copy preserve_bindings and preserve_spec_constants from the context.
+  // We OR the arguments together with the arguments passed into the constructor
+  // since the arguments in the constructor are per-pass overrides
+  // (e.g. we shouldn't strip bindings or spec constants for HLSL legalization).
+  preserve_bindings_ = preserve_bindings_ || context()->preserve_bindings();
+  preserve_spec_constants_ = preserve_spec_constants_ || context()->preserve_spec_constants();
```

Comment:
I don't like having two ways of doing the same thing.  Either have the options to ADCE or have a flag in the context.  I think the flag in the context is a good idea.